snuba/web/rpc/v1/trace_item_attribute_values.py

```diff
@@ -19,97 +19,114 @@
-                ),
-            )
-        )
-    else:
-        return (
-            base_conditions_and(
-                request.meta,
-                f.equals(column("attr_key"), literal(request.key.name)),
-                convert_filter_offset(request.page_token.filter_offset),
-            )
-            if request.page_token.HasField("filter_offset")
-            else base_conditions_and(
-                request.meta,
-                f.equals(column("attr_key"), literal(request.key.name)),
+def _build_conditions(request: TraceItemAttributeValuesRequest) -> Expression:
+    attribute_key = attribute_key_to_expression_eap_items(request.key)
+
+    conditions: list[Expression] = [
+        f.has(column("_hash_map_string"), request.key.name),
+    ]
+    if request.meta.trace_item_type:
+        conditions.append(f.equals(column("item_type"), request.meta.trace_item_type))
+
+    if request.value_substring_match:
+        conditions.append(
+            f.like(
+                attribute_key,
+                f"%{request.value_substring_match}%",
             )
         )
 
+    return base_conditions_and(request.meta, *conditions)
+
 
-def _build_query(request: TraceItemAttributeValuesRequest) -> Query:
+def _build_query(
+    request: TraceItemAttributeValuesRequest,
+) -> CompositeQuery[LogicalDataSource]:
+    """Example query:
+
+
+    SELECT distinct(attr_value) FROM
+    (
+        SELECT attributes_string_38['sentry.description'] as attr_value
+        FROM eap_items_1_dist
+        WHERE
+        has(_hash_map_string_38, cityHash64('sentry.description'))
+        AND attributes_string_38['sentry.description'] LIKE '%django.middleware%'
+        AND project_id = 1 AND organization_id=1 AND item_type=1
+        AND less(timestamp, toDateTime(1741910400))
+        AND greaterOrEquals(timestamp, toDateTime(1741651200))
+        ORDER BY attr_value
+        LIMIT 10000
+    ) ORDER BY attr_value LIMIT 1000
+
+
+    This query will match the first 10000 occurrences of an attribute value and then deduplicate them,
+    this gives a large speedup to the query at the cost of ordering and paginating all values
+    """
     if request.limit > 1000:
         raise BadSnubaRPCRequestException("Limit can be at most 1000")
 
+    entity_key = EntityKey("eap_items")
     entity = Entity(
-        key=EntityKey("spans_str_attrs"),
-        schema=get_entity(EntityKey("spans_str_attrs")).get_data_model(),
+        key=entity_key,
+        schema=get_entity(entity_key).get_data_model(),
         sample=None,
     )
-
-    truncate_request_meta_to_day(request.meta)
-
-    res = Query(
+    attr_value = attribute_key_to_expression_eap_items(request.key)
+    assert attr_value.alias
+    inner_query = Query(
         from_clause=entity,
+        selected_columns=[
+            SelectedExpression(name=attr_value.alias, expression=attr_value)
+        ],
+        condition=_build_conditions(request),
+        offset=request.page_token.offset,
```

Comment:
The `offset` is calculated in the `inner_query` and also on L120 in calculating `res`. Is this intentional? Wondering if the pagination is needed at both inner and outer levels and so should inner `offset` be removed?

Reply:
offset shouldn't be here at all actually, good catch